monadic-rx/shared/src/main/scala/mhtml/Optimizer.md

```diff
@@ -0,0 +1,191 @@
+    // numbers =>    0       0   3   4        5   6 ...
+    // odd1    =>       •-1      3            5     ...
+    // odd2    =>                       •-1   5     ...    
+    ```
+    
+We see that `odd1` emits -1, 3, 5 while `even2` emits -1, 5, so not everything can be 
+shared between the two streams. In other words, we cannot simply implement `odd2` by 
+"hooking" into `odd1` upon registration, since this would instead give the emission 
+sequence 3, 5. But we could do that after the 5 is emitted! 
+
+A similar argument can be used to show that merge is not always shareable but only 
+tail-shareable.
+    
+//TODO:    
+* `merge`
+    
+    ```scala
+    val r1: Rx[Int]
+    val r2: Rx[Int]
+    val merged: Rx[Int] = r1.merge(r2)
+    // r1      => 0    8     3 ...
+    // r2      => 1      4 3   ...
+    // merged1 => •0 1 8 4 3 3 ...
+    // merged2 =>   •1 8 4 3 3 ...
+    ```
+
+
+* `fold`
+
+    ```scala
+    val numbers: Rx[Int]
+    val folded: Rx[Int] = numbers.foldp(0)(_ + _)
+    // numbers  => 1  2 1  1 3 ...
+    // folded1  => •1 3 4  5 8 ...
+    // folded2  =>      •1 2 5 ...
+    ```
+
+
+(TODO: I think flatMap can be made more precise by saying that it "inherits" the 
+shareability of whatever is returned by its function. For instance, flatMapping on something that always returns a Rx made of map and zip results in something always shareable.)
+
+* `flatMap` basic case (always shareable)
+
+    ```scala
+    val numbers: Rx[Int]
+    val evens: Rx[Int]
+    val odds: Rx[Int]
+    val flip: Rx[Int] = numbers.flatMap(x => if x % 2 == 0 evens else odds)
+    // numbers   => 1  2 1  4 3 ...
+    // evens     => 0  2 4  6 8 ...
+    // odds      => 1  3 5  7 9 ...
+    // flip1     => •1 2 5  6 9 ...
+    // flip2     =>      •5 6 9 ...
+    ```
+
+//TODO: other `flatMap` cases
+
+It's then interesting to think how this shareability properties compose. 
+I have the intuition than if a Rx is only composed of always shareable parts it 
+is itself always shareable, while containing any never shareable component makes 
+the entire Rx non shareable. It gets funny when mixing always shareable and 
+tail shareable components, it seams that the result is not just tail shareable, 
+but something stronger where each of the intermediate streams have emitted at
+least one element. As an example, consider 
+`val out = rx1.dropIf(_ % 2 == 0)(-1).merge(rx2.dropIf(_ % 2 == 0)(-1))`. 
+It is possible to have out emit several values without reaching shareability, 
+or to have each "leaf" Var to emit several values without reaching shareability 
+(the diagrams get a bit messy to draw, but I have have them on a white board .
+
+
+## Algorithmic Considerations
+
+**Question**: since we already have asserted that:
+> When combined with a `fold` operator, `flatMap` is problematic: 
+it either leaks memory or breaks referential transparency.
+
+should we detect this and either warn or throw an error?
```

Comment:
This is not the case with mhtml flatMap's due to it's weird "registration" semantic, that is, the current implementation is both memory safe and RT.

What makes this possible is that flatMap is defined such that when it's upstream Rx updates it *cancels* every that did before, call it's `f: A => Rx[B]` with the new `A` and *registers* to the newly returned `Rx[B]`.

The consequence is that if you do something like the following:

```scala
unstream.flatMap(x => other.fold()())
```

Every-time `unstream` updates, a fresh new `fold` is started from on top of the current value of `other`. I hope this makes it obvious why `flatMap` inherits the shareability of whatever its `f` returns :smile: